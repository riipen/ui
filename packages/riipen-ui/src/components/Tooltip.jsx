import PropTypes from "prop-types";
import React from "react";
import css from "styled-jsx/css";

import ThemeContext from "../styles/ThemeContext";

import withClasses from "../utils/withClasses";

import Popover from "./Popover";

class Tooltip extends React.Component {
  static displayName = "Tooltip";

  static propTypes = {
    /**
     * The content to trigger the tooltip with.
     */
    children: PropTypes.node,

    /**
     * Any additional classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Color of tooltip.
     */
    color: PropTypes.oneOf([
      "default",
      "white",
      "positive",
      "negative",
      "warning"
    ]),

    /**
     * Whether tooltip should display on click.
     */
    click: PropTypes.bool,

    /**
     * The component used for the root node.
     * Either a string to use a DOM element or a component.
     */
    component: PropTypes.elementType,

    /**
     * Function to call on tooltip keydown.
     */
    onKeyDown: PropTypes.func,

    /**
     * Whether tooltip should display on hover.
     */
    hover: PropTypes.bool,

    /**
     * How the open/close state will be determined:
     * [Default] false: by this component's state,
     * true: by this component's props.
     */
    isControlledByProps: PropTypes.bool,

    /**
     * Function to call on tooltip close.
     */
    onClose: PropTypes.func,

    /**
     * Function to call on tooltip open.
     */
    onOpen: PropTypes.func,

    /**
     * Whether the popover should be open/shown.
     * Only used when props.isControlledByProps is true
     * [Default] false
     */
    open: PropTypes.bool,

    /**
     * Where to display the tooltip in relation to element.
     */
    position: PropTypes.oneOf([
      "top-right",
      "top-center",
      "top-left",
      "center-right",
      "center-left",
      "bottom-right",
      "bottom-center",
      "bottom-left"
    ]),

    /**
     * The size of the tooltip.
     */
    size: PropTypes.oneOf(["small", "medium"]),

    /**
     * The tooltip content to display.
     */
    tooltip: PropTypes.node
  };

  static defaultProps = {
    classes: [],
    click: false,
    color: "default",
    component: "div",
    hover: true,
    isControlledByProps: false,
    open: false,
    position: "bottom-center",
    size: "small"
  };

  constructor() {
    super();
    this.tooltipRootRef = React.createRef();
    this.state = {
      open: false
    };
  }

  getLinkedStyles = () => {
    const theme = this.context;

    return css.resolve`
      .popover {
        border-radius: 2px;
        color: ${theme.palette.common.white};
      }

      @keyframes fade {
        from {
          opacity: 0;
        }

        to: {
          opacity: 1;
        }
      }

      .popover.show {
        animation: fade ${theme.transitions.duration.short}ms
          ${theme.transitions.easing.easeIn};
      }

      /* Sizes */

      .popover.small {
        padding: ${theme.spacing(1)}px ${theme.spacing(4)}px;
      }

      .popover.medium {
        padding: ${theme.spacing(4)}px;
      }

      /* Colors */

      .popover.default {
        --after-color: ${theme.palette.grey[600]};

        --before-color: ${theme.palette.grey[600]};

        background: ${theme.palette.grey[600]};
        border: 1px solid ${theme.palette.grey[600]};
      }

      .popover.negative {
        --after-color: ${theme.palette.negative.main};

        --before-color: ${theme.palette.negative.main};

        background: ${theme.palette.negative.main};
        border: 1px solid ${theme.palette.negative.main};
      }

      .popover.positive {
        --after-color: ${theme.palette.positive.main};

        --before-color: ${theme.palette.positive.main};

        background: ${theme.palette.positive.main};
        border: 1px solid ${theme.palette.positive.main};
      }

      .popover.warning {
        --after-color: ${theme.palette.warning.main};

        --before-color: ${theme.palette.warning.main};

        background: ${theme.palette.warning.main};
        border: 1px solid ${theme.palette.warning.main};
      }

      .popover.white {
        --after-color: ${theme.palette.common.white};

        --before-color: ${theme.palette.grey[300]};

        background: ${theme.palette.common.white};
        border: 1px solid ${theme.palette.grey[300]};
        color: ${theme.palette.text.primary};
      }

      /* Positions */

      /* Tooltip arrow border */
      .popover::before {
        border-color: transparent;
        border-style: solid;
        border-width: ${theme.spacing(2)}px;
        content: "";
        position: absolute;
        white-space: normal;
      }

      /* Tooltip arrow inside */
      .popover::after {
        border-color: transparent;
        border-style: solid;
        border-width: calc(${theme.spacing(2)}px - 1px);
        content: "";
        position: absolute;
        white-space: normal;
      }

      .popover.bottom {
        margin-top: ${theme.spacing(3)}px;
      }

      .popover.bottom::before {
        margin-top: ${theme.spacing(-4)}px;
        top: 0;
      }

      .popover.bottom::after {
        margin-top: -17px;
        top: 0;
      }

      .popover.bottom-right {
        border-top-left-radius: 0;
      }

      /* Tooltip arrow */
      .popover.bottom-right::before {
        border-bottom-color: var(--before-color);
        border-left: none;
        left: 0;
        margin-left: -1px;
      }

      .popover.bottom-right::after {
        border-bottom-color: var(--after-color);
        border-left: none;
        left: 0;
      }

      .popover.bottom-center::before {
        border-bottom-color: var(--before-color);
        left: calc(50% - (${theme.spacing(2)}px));
      }

      .popover.bottom-center::after {
        border-bottom-color: var(--after-color);
        border-width: ${theme.spacing(2)}px;
        left: calc(50% - ${theme.spacing(2)}px);
        margin-top: calc(${theme.spacing(-4)}px + 1px);
      }

      .popover.bottom-left {
        border-top-right-radius: 0;
      }

      .popover.bottom-left::before {
        border-bottom-color: var(--before-color);
        border-right: none;
        margin-right: -1px;
        right: 0;
      }

      .popover.bottom-left::after {
        border-bottom-color: var(--after-color);
        border-right: none;
        right: 0;
      }

      .popover.center::before {
        top: calc(50% - ${theme.spacing(2)}px);
      }

      .popover.center::after {
        top: calc(50% - ${theme.spacing(2)}px);
      }

      .popover.center-right {
        margin-left: ${theme.spacing(3)}px;
      }

      .popover.center-right::before {
        border-right-color: var(--before-color);
        left: 0;
        margin-left: ${theme.spacing(-4)}px;
      }

      .popover.center-right::after {
        border-right-color: var(--after-color);
        border-width: ${theme.spacing(2)}px;
        left: 0;
        margin-left: calc(${theme.spacing(-4)}px + 1px);
      }

      .popover.center-left {
        margin-left: ${theme.spacing(-3)}px;
      }

      .popover.center-left::before {
        border-left-color: var(--before-color);
        margin-right: ${theme.spacing(-4)}px;
        right: 0;
      }

      .popover.center-left::after {
        border-left-color: var(--after-color);
        border-width: ${theme.spacing(2)}px;
        margin-right: calc(${theme.spacing(-4)}px + 1px);
        right: 0;
      }

      .popover.top {
        margin-top: ${theme.spacing(-3)}px;
      }

      .popover.top::before {
        bottom: 0;
        margin-bottom: ${theme.spacing(-4)}px;
      }

      .popover.top::after {
        bottom: 0;
        margin-bottom: -17px;
      }

      .popover.top-right {
        border-bottom-left-radius: 0;
      }

      .popover.top-right::before {
        border-left: none;
        border-top-color: var(--before-color);
        left: 0;
        margin-left: -1px;
      }

      .popover.top-right::after {
        border-left: none;
        border-top-color: var(--after-color);
        left: 0;
      }

      .popover.top-center::before {
        border-top-color: var(--before-color);
        left: calc(50% - ${theme.spacing(2)}px);
      }

      .popover.top-center::after {
        border-top-color: var(--after-color);
        border-width: ${theme.spacing(2)}px;
        left: calc(50% - ${theme.spacing(2)}px);
        margin-bottom: calc(${theme.spacing(-4)}px + 1px);
      }

      .popover.top-left {
        border-bottom-right-radius: 0;
      }

      .popover.top-left::before {
        border-right: none;
        border-top-color: var(--before-color);
        border-width: ${theme.spacing(2)}px;
        margin-right: -1px;
        right: 0;
      }

      .popover.top-left::after {
        border-right: none;
        border-top-color: var(--after-color);
        right: 0;
      }
    `;
  };

  /**
   * Gets the opposing direction of direction1 and direction2.
   * @param {string} direction1 - first direction
   * @param {string} direction2 - opposing direction of direction1
   * @param {string} value - value to compare
   * @returns {string} Opposing direciton, or just value if directions do not oppose.
   */
  getOppositeDirection = (direction1, direction2, value) => {
    if (value === direction1) {
      return direction2;
    } else if (value === direction2) {
      return direction1;
    }
    return value;
  };

  blur = () => {
    // Unfocus if it is still active
    const activeElement = document.activeElement;
    if (activeElement) {
      activeElement.blur();
    }
  };

  clickCallback = () => {
    const { open } = this.state;

    const callback = !open ? this.props.onOpen : this.props.onClose;

    this.blur();

    this.setState(
      {
        open: !open
      },
      callback
    );
  };

  static contextType = ThemeContext;

  handleOpen = () => {
    this.setState({ open: true }, this.props.onOpen);
  };

  handleClose = () => {
    this.blur();

    this.setState(
      {
        open: false
      },
      this.props.onClose
    );
  };

  renderPopover = () => {
    const {
      classes,
      color,
      onKeyDown,
      isControlledByProps,
      position,
      size,
      tooltip,
      ...other
    } = this.props;

    const { open } = isControlledByProps ? this.props : this.state;

    const linkedStyles = this.getLinkedStyles();

    const [vertical, horizontal] = position.split("-");

    const contentVertical = this.getOppositeDirection(
      "top",
      "bottom",
      vertical
    );
    const contentHorizontal = this.getOppositeDirection(
      "left",
      "right",
      horizontal
    );

    return (
      <Popover
        classes={classes.concat([
          linkedStyles.className,
          "popover",
          color,
          position,
          vertical,
          size,
          open && "show"
        ])}
        anchorPosition={{
          horizontal,
          vertical
        }}
        contentPosition={{
          horizontal: contentHorizontal,
          vertical: contentVertical
        }}
        onKeyDown={onKeyDown}
        anchorEl={this.tooltipRootRef.current}
        isOpen={open}
        keepOnScreen
        lockScroll={false}
        onClose={this.handleClose}
        {...other}
      >
        <React.Fragment>{tooltip}</React.Fragment>
      </Popover>
    );
  };
  render() {
    const { children, click, hover, component: Component } = this.props;

    const linkedStyles = this.getLinkedStyles();

    return (
      <React.Fragment>
        <Component
          ref={this.tooltipRootRef}
          onClick={click ? this.clickCallback : undefined}
          onMouseEnter={hover ? this.handleOpen : undefined}
          onMouseLeave={hover ? this.handleClose : undefined}
        >
          {children}
        </Component>
        {this.renderPopover()}
        {linkedStyles.styles}
      </React.Fragment>
    );
  }
}

export default withClasses(Tooltip);
