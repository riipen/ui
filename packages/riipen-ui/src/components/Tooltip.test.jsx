import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Tooltip from "./Tooltip.jsx";

describe("<Tooltip>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Tooltip />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Tooltip />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("children prop", () => {
    it("renders the children", () => {
      const children = "test";
      const wrapper = mount(<Tooltip>{children}</Tooltip>);

      expect(wrapper.text()).toEqual(children);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the popover node", () => {
      const classes = ["classOne"];
      const wrapper = mount(<Tooltip classes={classes} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().classes
      ).toContain(classes[0]);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-tooltip"];
      const wrapper = mount(<Tooltip />);

      expect(
        wrapper
          .find("Tooltip")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });
  });

  describe("click prop", () => {
    it("invokes click when click is true", () => {
      const click = true;
      const onOpen = jest.fn();
      const wrapper = mount(<Tooltip click={click} onOpen={onOpen} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("click");

      expect(onOpen).toHaveBeenCalledTimes(1);
    });
  });

  describe("color prop", () => {
    it("sets tooltip color with given valid custom color prop", () => {
      const color = "positive";
      const wrapper = mount(<Tooltip color={color} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().classes
      ).toContain(color);
    });

    it("gives an error given invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const color = "bad";
      mount(<Tooltip color={color} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("component prop", () => {
    it("sets custom component", () => {
      const component = "span";
      const wrapper = mount(<Tooltip component={component} />);

      expect(
        wrapper
          .find("Tooltip")
          .childAt(0)
          .name()
      ).toEqual(component);
    });

    it("sets component type to div when component prop is not provided", () => {
      const wrapper = mount(<Tooltip />);

      expect(
        wrapper
          .find("Tooltip")
          .childAt(0)
          .name()
      ).toEqual("div");
    });
  });

  describe("hover prop", () => {
    it("opens tooltip on mouse enter when hover is true", () => {
      const wrapper = mount(<Tooltip />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");

      expect(wrapper.find("Popover").props().isOpen).toBe(true);
    });

    it("closes tooltip on mouse leave when hover is true", () => {
      const wrapper = mount(<Tooltip />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");
      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseLeave");

      expect(wrapper.find("Popover").props().isOpen).toBe(false);
    });

    it("does not open tooltip on mouse enter when hover is false", () => {
      const hover = false;
      const wrapper = mount(<Tooltip hover={hover} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");

      expect(wrapper.find("Popover").props().isOpen).toBe(false);
    });

    it("does not close tooltip on mouse leave when hover is false", () => {
      const hover = false;
      const click = true;
      const wrapper = mount(<Tooltip click={click} hover={hover} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("click");
      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseLeave");

      expect(wrapper.find("Popover").props().isOpen).toBe(true);
    });
  });

  describe("isControlledByProps prop", () => {
    it("does not open tooltip on mouse enter when onOpen not provided", () => {
      const isControlledByProps = true;
      const open = false;
      const wrapper = mount(
        <Tooltip isControlledByProps={isControlledByProps} open={open} />
      );

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");

      expect(wrapper.find("Popover").props().isOpen).toBe(false);
    });

    it("does not close tooltip on mouse leave when onOpen not provided", () => {
      const isControlledByProps = true;
      const open = true;
      const wrapper = mount(
        <Tooltip isControlledByProps={isControlledByProps} open={open} />
      );

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseLeave");

      expect(wrapper.find("Popover").props().isOpen).toBe(true);
    });
  });

  describe("keepOpenOnMouseLeave prop", () => {
    it("does not close tooltip on mouse leave", () => {
      const keepOpenOnMouseLeave = true;
      const wrapper = mount(
        <Tooltip keepOpenOnMouseLeave={keepOpenOnMouseLeave} />
      );

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");
      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseLeave");

      expect(wrapper.find("Popover").props().isOpen).toBe(true);
    });
  });

  describe("onClose prop", () => {
    it("invokes onOpen prop on mouse leave", () => {
      const onClose = jest.fn();
      const wrapper = mount(<Tooltip onClose={onClose} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseLeave");

      expect(onClose).toHaveBeenCalledTimes(1);
    });

    it("invokes onClose prop on click when click is true", () => {
      const click = true;
      const onClose = jest.fn();
      const wrapper = mount(<Tooltip click={click} onClose={onClose} />);

      // open and close tooltip
      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("click");
      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("click");

      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe("onKeyDown prop", () => {
    it("passes onKeyDown to Popover", () => {
      const onKeyDown = jest.fn();
      const wrapper = mount(<Tooltip onKeyDown={onKeyDown} />);

      expect(wrapper.find("Popover").props().onKeyDown).toEqual(onKeyDown);
    });
  });

  describe("onOpen prop", () => {
    it("invokes onOpen prop on mouse enter", () => {
      const onOpen = jest.fn();
      const wrapper = mount(<Tooltip onOpen={onOpen} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("mouseEnter");

      expect(onOpen).toHaveBeenCalledTimes(1);
    });

    it("invokes onOpen prop on click when click is true", () => {
      const click = true;
      const onOpen = jest.fn();
      const wrapper = mount(<Tooltip click={click} onOpen={onOpen} />);

      wrapper
        .find("Tooltip")
        .childAt(0)
        .simulate("click");

      expect(onOpen).toHaveBeenCalledTimes(1);
    });
  });

  describe("open prop", () => {
    it("passes open to Popover as true when provided", () => {
      const isControlledByProps = true;
      const open = true;
      const tooltip = "test";

      const wrapper = mount(
        <Tooltip
          isControlledByProps={isControlledByProps}
          open={open}
          tooltip={tooltip}
        />
      );

      expect(wrapper.find("Popover").props().isOpen).toBe(true);
    });
  });

  describe("position prop", () => {
    it("passes correct position class to Popover", () => {
      const position = "top-left";
      const wrapper = mount(<Tooltip position={position} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().classes
      ).toContain(position);
    });

    it("passes correct contentPosition prop to Popover", () => {
      const position = "top-left";
      const wrapper = mount(<Tooltip position={position} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().contentPosition
      ).toStrictEqual({ horizontal: "right", vertical: "bottom" });
    });
  });

  describe("size prop", () => {
    it("sets tooltip size with given valid custom size prop", () => {
      const size = "small";
      const wrapper = mount(<Tooltip size={size} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().classes
      ).toContain(size);
    });

    it("gives an error given invalid size", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const size = "sm";
      mount(<Tooltip size={size} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });

  describe("tooltip prop", () => {
    it("renders tooltip inside Popover", () => {
      const tooltip = "test";
      const wrapper = mount(<Tooltip tooltip={tooltip} />);

      expect(
        wrapper
          .find("Tooltip")
          .find("Popover")
          .props().children.props.children
      ).toBe(tooltip);
    });
  });
});
