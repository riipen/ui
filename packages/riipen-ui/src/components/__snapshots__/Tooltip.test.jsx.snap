// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip> renders correct snapshot 1`] = `
<withClasses(Tooltip)>
  <Tooltip
    classes={
      Array [
        "riipen",
        "riipen-tooltip",
      ]
    }
    click={false}
    color="default"
    component="div"
    hover={true}
    isControlledByProps={false}
    keepOpenOnMouseLeave={false}
    open={false}
    position="bottom-center"
    size="small"
  >
    <div
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    />
    <withClasses(Popover)
      anchorEl={null}
      anchorPosition={
        Object {
          "horizontal": "center",
          "vertical": "bottom",
        }
      }
      classes={
        Array [
          "riipen",
          "riipen-tooltip",
          "jsx-3429463278",
          "popover",
          "default",
          "bottom-center",
          "bottom",
          "small",
          false,
        ]
      }
      contentPosition={
        Object {
          "horizontal": "center",
          "vertical": "top",
        }
      }
      isOpen={false}
      keepOnScreen={true}
      lockScroll={false}
      onClose={[Function]}
    >
      <Popover
        anchorEl={null}
        anchorPosition={
          Object {
            "horizontal": "center",
            "vertical": "bottom",
          }
        }
        classes={
          Array [
            "riipen",
            "riipen-popover",
            "riipen",
            "riipen-tooltip",
            "jsx-3429463278",
            "popover",
            "default",
            "bottom-center",
            "bottom",
            "small",
            false,
          ]
        }
        closeOnScrolled={true}
        component="span"
        contentPosition={
          Object {
            "horizontal": "center",
            "vertical": "top",
          }
        }
        fullWidth={false}
        isOpen={false}
        keepOnScreen={true}
        lockScroll={false}
        marginThreshold={16}
        onClose={[Function]}
        styles={Object {}}
      >
        <JSXStyle
          dynamic={
            Array [
              "#fff",
              "4px",
              "0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12)",
              "calc(100% - 32px)",
              "",
              3,
            ]
          }
          id="631859544"
        />
      </Popover>
    </withClasses(Popover)>
    <JSXStyle
      dynamic={
        Array [
          "#fff",
          250,
          "cubic-bezier(0.4, 0, 1, 1)",
          5,
          20,
          20,
          "#747474",
          "#747474",
          "#747474",
          "#747474",
          "#e95353",
          "#e95353",
          "#e95353",
          "#e95353",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#4cc88d",
          "#febb46",
          "#febb46",
          "#febb46",
          "#febb46",
          "#fff",
          "#dddddd",
          "#fff",
          "#dddddd",
          "#373737",
          10,
          10,
          15,
          -20,
          10,
          10,
          10,
          -20,
          10,
          10,
          15,
          -20,
          10,
          -20,
          -15,
          -20,
          10,
          -20,
          -15,
          -20,
          10,
          10,
          10,
          -20,
          10,
        ]
      }
      id="4063941441"
    />
  </Tooltip>
</withClasses(Tooltip)>
`;
